### PR TITLE
Initial commit for Oracle Cloud Infrastructure modules

### PR DESCRIPTION
**Pull request by nalsaber**
##### SUMMARY
Initial commit for Oracle cloud infrastructure

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
Oracle cloud infrastructure Manage Virtual Cloud Networks(VCN) module

##### ADDITIONAL INFORMATION
Files Submitted:
lib/ansible/module_utils/oracle/__init__.py
lib/ansible/module_utils/oracle/oci_logging.yaml
lib/ansible/module_utils/oracle/oci_utils.py
lib/ansible/modules/cloud/oracle/__init__.py
lib/ansible/modules/cloud/oracle/oci_vcn.py
lib/ansible/plugins/doc_fragments/oracle.py
lib/ansible/plugins/doc_fragments/oracle_creatable_resource.py
lib/ansible/plugins/doc_fragments/oracle_display_name_option.py
lib/ansible/plugins/doc_fragments/oracle_name_option.py
lib/ansible/plugins/doc_fragments/oracle_tags.py
lib/ansible/plugins/doc_fragments/oracle_wait_options.py

### CODE REVIEW

File: lib/ansible/plugins/doc_fragments/oracle.py
Comment (gundalow) [inline lib/ansible/plugins/doc_fragments/oracle.py]: `required: false` is the default so can be removed
```suggestion
```

File: lib/ansible/plugins/doc_fragments/oracle.py
Comment (gundalow) [inline lib/ansible/plugins/doc_fragments/oracle.py:20]: To improve the quality of the docs please add `type:` for all options in docs_fragments and the modules themselves.
```suggestion
            default: "DEFAULT"
            type: str
```

File: lib/ansible/modules/cloud/oracle/oci_vcn.py
Comment (gundalow) [inline lib/ansible/modules/cloud/oracle/oci_vcn.py]: In Ansible 2.8 we added a helpter method method for this https://github.com/ansible/ansible/pull/51787
```suggestion
        module.fail_json(msg=missing_required_lib("oci"))
```

File: lib/ansible/plugins/doc_fragments/oracle.py
Comment (nalsaber) [reply lib/ansible/plugins/doc_fragments/oracle.py]: removed all references

File: lib/ansible/plugins/doc_fragments/oracle.py
Comment (nalsaber) [reply lib/ansible/plugins/doc_fragments/oracle.py:20]: added the type

File: lib/ansible/modules/cloud/oracle/oci_vcn.py
Comment (nalsaber) [reply lib/ansible/modules/cloud/oracle/oci_vcn.py]: done

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (gundalow) [inline lib/ansible/module_utils/oracle/oci_utils.py:96]: Do you need this? 
Could you just see if api_key is set or not?

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (gundalow) [inline lib/ansible/module_utils/oracle/oci_utils.py:89]: Please remove `required=False` that's the default

File: lib/ansible/modules/cloud/oracle/oci_vcn.py
Comment (gundalow) [inline lib/ansible/modules/cloud/oracle/oci_vcn.py]: I think this should be the name of the library

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (gundalow) [inline lib/ansible/module_utils/oracle/oci_utils.py]: The aliases should be available via the standard name. I'm not sure what this function is for

ie
```python
display_name=dict(type="str", required=False, aliases=["name"]),
foo = module.params["display_name"]
# foo would be set to the what's passed in as `display_name` or `name`

```

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (nalsaber) [reply lib/ansible/module_utils/oracle/oci_utils.py:89]: sure

File: lib/ansible/modules/cloud/oracle/oci_vcn.py
Comment (nalsaber) [reply lib/ansible/modules/cloud/oracle/oci_vcn.py]: The library is called oci-python-sdk

Comment (gundalow) [inline lib/ansible/modules/cloud/oracle/oci_vcn.py]:
```suggestion
        module.fail_json(msg=missing_required_lib("oci-python-sdk"))
```

Comment (felixfontein) [reply lib/ansible/modules/cloud/oracle/oci_vcn.py]: I think it is just called `oci`: https://pypi.org/project/oci/ At least https://pypi.org/project/oci-python-sdk/ does not exist.

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (rohitChaware) [reply lib/ansible/module_utils/oracle/oci_utils.py]: The function `has_user_provided_value_for_option()` figures out if an user has actually provided value for `option` using standard name or alias. This helps in preparing the list of attributes of a resource which need to be updated in the function `get_attr_to_update()`.

`module.params` has all the standard options set to some value(user provided, default or `None`) irrespective of whether user provided it or not.

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (rohitChaware) [reply lib/ansible/module_utils/oracle/oci_utils.py:96]: There may be a case when `api_key` is set but still an user wants to authenticate using instance principal by setting `auth_type` to `instance_principal`.

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (felixfontein) [reply lib/ansible/module_utils/oracle/oci_utils.py]: But *why* to do you need to figure out whether the option was provided via the main name or an alias? No other module makes any distinction by which name the option values are passed AFAIK. So why break with this convention?

Comment (rohitChaware) [reply lib/ansible/module_utils/oracle/oci_utils.py]: Thanks @felixfontein for reviewing this closely.

We have generalized the update utility[`check_and_update_resource()`] which is used by all the OCI Ansible modules for update scenarios.

To have a consistent design, the names of attributes of the update model of a resource become 'options with main name or alias' in the Ansible module for the resource.

For example, attributes: `defined_tags, display_name, freeform_tags` of the update model: https://oracle-cloud-infrastructure-python-sdk.readthedocs.io/en/latest/api/core/models/oci.core.models.UpdateVcnDetails.html are options in `oci_vcn` Ansible module and attribute:`display_name` could be alias for an option:`name`.

Then while iterating over the update model of a resource to figure out if the value of an attribute is actually provided by the user, we look for this attribute name in main names or aliases. If we don’t look in aliases, we may miss out as in the following case:

`display_name` could be alias for an option having standard name as `name`. While iterating on the update model to check if value of attribute:`display_name` is provided, we have to look for it in aliases too if user has used option:`name` to specify the value since `params["name"]` will be set and not `params["display_name"]`.

Comment (felixfontein) [reply lib/ansible/module_utils/oracle/oci_utils.py]: If `name` is an alias of `display_name`, and the user sets `name: xxx`, `params['display_name']` will be `'xxx'`. Ansible does that internally so no module has to do that.

Comment (rohitChaware) [reply lib/ansible/module_utils/oracle/oci_utils.py]: Agree and it is true for `module.params`.

In `has_user_provided_value_for_option()`, we are processing return value of `_load_params()`, which is raw user input dictionary and has not gone through the default value substitution and alias processing. 

We need raw user input so that we can distinguish between cases when (a) a user explicitly assigns a value to an option and (b) a value gets assigned because of default value for the option.

Once we know what options are explicitly specified by a user, we update only those attributes.

Comment (felixfontein) [reply lib/ansible/module_utils/oracle/oci_utils.py]: I think the standard way to do this (i.e. what all other modules do) is to not specify a default value for that option, and use `None` as "was not specified by user". I don't know how the core team sees this. If modules should be allowed to know where a value comes from (user, default value, environment, ...), I think there should be a common mechanism which can be used by all modules.

File: lib/ansible/modules/cloud/oracle/oci_vcn.py
Comment (nalsaber) [reply lib/ansible/modules/cloud/oracle/oci_vcn.py]: I will rename it to 'oci'

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (nalsaber) [reply lib/ansible/module_utils/oracle/oci_utils.py]: @sivel @samdoran hey guys do you have any input on this?

Comment (samdoran) [reply lib/ansible/module_utils/oracle/oci_utils.py]: This feels a lot like reimplementing argument spec validation methods inside a module. I would advise against this.

> We need raw user input so that we can distinguish between cases when (a) a user explicitly assigns a value to an option and (b) a value gets assigned because of default value for the option.

I don't fully understand this need. The module should not care if a user provided the value or if it came from the default setting in the argument spec. It should only care about the value. Treating the value differently based on where it was set, particularly if the provided value is the same as the default value, seems overly complicated.

Comment (rohitChaware) [reply lib/ansible/module_utils/oracle/oci_utils.py]: Thanks @felixfontein and @samdoran for your inputs. 

We have a few module options in some OCI Ansible modules which are common to create and update operations. These options have defaults during create operation(as API puts in default value during create operation if nothing is specified) but don't have any defaults for update operation. 

As suggested under `options:`->`default:` on https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fields, we preferred mentioning default values for these options using the `default` construct instead of adding to option's documentation about defaults.

So during update operation when a user hasn't specified a value for such an option, the option gets substituted by the default value. Thus we have to check if the user provided the value or it came from the default setting in argument spec before we actually proceed for the update operation. If we don't make this check, we will end up updating the attribute of the resource even if the user didn't intend to.

Comment (nalsaber) [reply lib/ansible/module_utils/oracle/oci_utils.py]: function removed

**Comment by nalsaber** [comment]
@gundalow Please review this PR. We addressed all of your concerns. Thanks

File: lib/ansible/module_utils/oracle/oci_utils.py
Comment (felixfontein) [inline lib/ansible/module_utils/oracle/oci_utils.py:1610]: What do you need this function for? Ansible already handles aliases, there's no need to do this manually.

Comment (nalsaber) [reply lib/ansible/module_utils/oracle/oci_utils.py]: @felixfontein can you please clarify your comment? I don't see any aliases at the comment line.

Comment (felixfontein) [reply lib/ansible/module_utils/oracle/oci_utils.py:1610]: (GitHub is showing this with the wrong code snippet in the discussion overview. This comment is about the `_get_user_provided_value` function.)

Comment (nalsaber) [reply lib/ansible/module_utils/oracle/oci_utils.py:1610]: @felixfontein  The purpose for _get_user_provided_value function is to allow us to compare the attributes returned from the cloud with the attributes provided by the user. If the values are different, we send an update request to the cloud. If we remove this function we have to change the naming convention that we use in all modules.

Currently we use this naming convention 
name=dict(type=“str”, required=False, aliases=[“display_name”]), 
Without this function, we need to flip the naming convention to to match the SDK attributes as the main values:
display_name=dict(type=“str”, required=False, aliases=[“name”])

We prefer to keep the current convention because it makes the modules' attributes more user friendly and more consistent. As you see, we are already using Ansible aliasing and this function is not used to handle aliasing or to re-implement an exciting functionality.

Comment (felixfontein) [reply lib/ansible/module_utils/oracle/oci_utils.py:1610]: Sounds ok.

**Comment by nalsaber** [comment]
@gundalow: Can you please merge this PR if you don't have any more comments?

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
rebuild_merge

**Comment by felixfontein** [comment]
@nalsaber @rohitChaware congratulations!

**Comment by mattclay** [comment]
@nalsaber This PR (as well as follow-up PRs https://github.com/ansible/ansible/pull/55888 and https://github.com/ansible/ansible/pull/55889) have been reverted in the devel and stable-2.8 branches.

There are three main issues with these changes that need to be addressed, and then they can be submitted again:

1. The license for modules must be GPLv3+ only.
2. The license for module_utils must be BSD only.
3. Non-code files such as `lib/ansible/module_utils/oracle/oci_logging.yaml` cannot be included under the `lib` directory. Modules will not have access to these files. On a related note, modules and module_utils generally should not implement their own logging infrastructure.

**Comment by mattclay** [comment]
It has come to my attention (thanks @felixfontein) that we have a somewhat recent change in our licensing guidance: https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_in_groups.html#submitting-a-group-of-modules

> A module_util used only for a specific vendor’s hardware, provider, or service may be licensed under GPLv3+.

However, based on some internal discussions, it seems that was intended as an allowable exception requiring approval from the Ansible Core Team, rather than a standard licensing option. It's clearly not worded that way in the docs currently, so we're looking at correcting that.

Dual licensing, however, as was used for the modules and module_utils in this PR (using the Apache 2.0 license) is not one of the permitted licensing options. The easiest way to correct the license for re-submission of this PR would be to use GPLv3+ for the modules and BSD for the module_utils. If you see a need for using GPLv3+ on the module_utils, please add it to the [Public Ansible Project Meeting Agenda](https://github.com/ansible/community/issues/456) for discussion.

My apologies that this workflow has been less than ideal. We've identified issues with our documentation around licensing that definitely need some attention. I'm also working on expanding our automated testing to detect both the unwanted files and unacceptable licensing at PR time to hopefully avoid future situations like this.

**Pull request merged**
